### PR TITLE
F: Enhanced group attribute handling and group image support

### PR DESCRIPTION
**Pull request by guymaness2**
##### What does the change add or fix?

* Added new "VM" option to group attribute handling to set up DefinedTypes that allow multiple values. When used, the value provided for each group will be treated as a comma separated list of individual DefinedValues.
* Added support for importing group images to binary importer. Images are added as group attributes of FileType "Group Image". 

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

* Added new "VM" option to group attribute handling to set up DefinedTypes that allow multiple values. When used, the value provided for each group will be treated as a comma separated list of individual DefinedValues.
* Added support for importing group images to binary importer. Images are added as group attributes of FileType "Group Image". 

See respective rep Wiki pages for usage instructions

---------

##### Who reported, requested, or paid for the change?

Victory World Church

---------

##### Does this update or add options to the block UI?

none

---------

##### What files does it affect?

* Bulldozer.BinaryFile/Bulldozer.BinaryFile.csproj
* Bulldozer.BinaryFile/Maps/GroupImage.cs
* Bulldozer.CSV/CSVComponent.cs
* Bulldozer.CSV/Maps/Group.cs
* Bulldozer/App.config
* Bulldozer/Utility/AddMethods.cs
* Bulldozer/Utility/Structures.cs

---------

##### Is it a breaking change for other versions/clients?

none

### CODE REVIEW

File: Bulldozer.BinaryFile/Maps/GroupImage.cs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                    completedItems++;
+                    if ( completedItems % percentage < 1 )
+                    {
+                        var percentComplete = completedItems / percentage;
+                        ReportProgress( percentComplete, string.Format( "{0:N0} group image files imported ({1}% complete).", completedItems, percentComplete ) );
+                    }
+
+                    if ( completedItems % ReportingNumber < 1 )
+                    {
+                        SaveFiles(newFileList);
```
Comment (guymaness2) [inline Bulldozer.BinaryFile/Maps/GroupImage.cs:170]:
```suggestion
                        SaveFiles( newFileList );
```

File: Bulldozer.BinaryFile/Maps/GroupImage.cs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        /// <summary>
+        /// Saves the files.
+        /// </summary>
+        /// <param name="newFileList">The new file list.</param>
+        private static void SaveFiles(List<GroupDocumentKeys> newFileList)
+        {
+            var rockContext = new RockContext();
+            rockContext.WrapTransaction( () =>
+            {
+                rockContext.BinaryFiles.AddRange(newFileList.Select(f => f.File));
```
Comment (guymaness2) [inline Bulldozer.BinaryFile/Maps/GroupImage.cs:197]: add check to condition

File: Bulldozer/Bulldozer.csproj [hunk 3]
```diff
@@ -98,60 +98,60 @@
     </Reference>
     <Reference Include="Microsoft.SqlServer.ConnectionInfo, Version=12.0.0.0, Culture=neutral, PublicKeyToken=89845dcd8080cc91, processorArchitecture=MSIL">
       <Private>True</Private>
-      <HintPath>..\packages\MicrosoftSQLServerSMO2014Wmi.1.0.1\lib\net45\Microsoft.SqlServer.ConnectionInfo.dll</HintPath>
+	  <HintPath>..\packages\MicrosoftSQLServerSMO2014Wmi.1.0.1\lib\net45\Microsoft.SqlServer.ConnectionInfo.dll</HintPath>
```
Comment (guymaness2) [inline Bulldozer/Bulldozer.csproj:101]:
```suggestion
<HintPath>..\packages\MicrosoftSQLServerSMO2014Wmi.1.0.1\lib\net45\Microsoft.SqlServer.ConnectionInfo.dll</HintPath>
```

Comment (guymaness2) [inline Bulldozer/Bulldozer.csproj:101]:
```suggestion
	 <HintPath>..\packages\MicrosoftSQLServerSMO2014Wmi.1.0.1\lib\net45\Microsoft.SqlServer.ConnectionInfo.dll</HintPath>
```

File: Bulldozer/packages.config [hunk 4]
```diff
@@ -9,7 +9,7 @@
   <package id="EntityFramework" version="6.1.3" targetFramework="net451" />
   <package id="Fody" version="3.2.9" targetFramework="net452" developmentDependency="true" />
   <package id="LumenWorksCsvReader" version="4.0.0" targetFramework="net452" />
-  <package id="MicrosoftSQLServerSMO2014Wmi" version="1.0.1" targetFramework="net451" />
+  <package id="MicrosoftSQLServerSMO2014Wmi" version="1.0.1" targetFramework="net452" />
```
Comment (guymaness2) [inline Bulldozer/packages.config:12]:
```suggestion
  <package id="MicrosoftSQLServerSMO2014Wmi" version="1.0.1" targetFramework="net451" />
```

File: Bulldozer.BinaryFile/Maps/GroupImage.cs (hunk 2 above)
Comment (nateh777) [reply Bulldozer.BinaryFile/Maps/GroupImage.cs:197]: @guymaness2 did this get taken care of?

**Review by nateh777 — CHANGES_REQUESTED** [review]
I did not see one conversation get resolved yet, please review and let me know what the result was: https://github.com/KingdomFirst/Bulldozer/pull/51/files#r675040097

File: Bulldozer.BinaryFile/Maps/GroupImage.cs (hunk 2 above)
Comment (guymaness2) [reply Bulldozer.BinaryFile/Maps/GroupImage.cs:197]: need to copy null check from line 199 to this one as well.

**Review by nateh777 — APPROVED** [review]

**Pull request merged**